Fixed #27833 -- Allowed QuerySet.prefetch_related() to avoid the max query parameter limitation of SQLite.

raphaelm opened the pull request:
prefetch_related used to fail with SQLite when used with 1000 parent records

see https://code.djangoproject.com/ticket/27833

raphaelm commented:
now waiting for the tests to be run on all databases.

@mjtamlyn I'd be happy about an initial review.

mjtamlyn commented on tests/prefetch_related/tests.py:
```diff
@@ -256,6 +258,77 @@ def test_m2m_then_reverse_one_to_one_object_ids(self):
+            batches = int(ceil(float(len(self.authors)) / batch_size))
+        else:
+            batches = 1
+        # One query for Avatar.objects.all() and then one related fast delete for
```
`fast delete`?

raphaelm replied:
m( copied code from the deletion tests here

mjtamlyn commented on tests/prefetch_related/tests.py:
```diff
@@ -256,6 +258,77 @@ def test_m2m_then_reverse_one_to_one_object_ids(self):
+
+        cls.reader = Reader.objects.create(name='Amy')
+
+        for author, book in zip(cls.authors, cls.books):
```
This is going to run a *lot* of separate queries to the database isn't it? Is there any way to bulk to this, or create less data in child tables we dont' need?

raphaelm replied:
Unfortunately, on SQLite and MySQL bulk_insert won't populate the .pk fields, so yes, we need to refetch everything from the database again.

Also, yes, we could probably get rid of some data in the child tables. However, we should have data in the child tables for the first and the last authors, so we know the chaining of the chunks is done properly.

I'm gonna try that tomorrow.

raphaelm commented:
I improved the comments and made the test as performant as I believe they can be.

raphaelm commented:
Thanks @francoisfreitag for the extensive review,I'll try to rework in the next few days!

raphaelm commented:
> Is there a reason not to use connections[prefetcher.db] to get the batch_size instead of passing the connection object all the way from prefetch_related_objects to prefetch_one_level?

I didn't see that this attribute exists ;)

raphaelm commented:
>> Is there a reason not to use connections[prefetcher.db] to get the batch_size instead of passing the connection object all the way from prefetch_related_objects to prefetch_one_level?
> I didn't see that this attribute exists ;)

Because apparently, it doesn't in all cases ("AttributeError: 'GenericForeignKey' object has no attribute 'db'" etc.)

raphaelm commented on tests/prefetch_related/tests.py:
```diff
@@ -256,6 +258,85 @@ def test_m2m_then_reverse_one_to_one_object_ids(self):
+        Author.objects.bulk_create(authors)
+        # Re-fetch the authors as .bulk_create() will not populate their .pk attribute
+        # on SQLite
+        cls.authors = list(Author.objects.all())
```
Will do

raphaelm commented on tests/prefetch_related/tests.py:
```diff
@@ -256,6 +258,85 @@ def test_m2m_then_reverse_one_to_one_object_ids(self):
+            # The current implementation uses a safety margin of 25 to
+            # cope with additional variables in the query.
+            batch_size = connection.features.max_query_params - 25
+            batches = int(ceil(float(len(self.authors)) / batch_size))
```
Good point

raphaelm commented on tests/prefetch_related/tests.py:
```diff
@@ -256,6 +258,85 @@ def test_m2m_then_reverse_one_to_one_object_ids(self):
+            self.assertEqual(1, len(result[0].books.all()))
+            self.assertEqual(1, len(result[-1].books.all()))
+
+    def test_recursive_lookups(self):
```
I saw the term for this being recursive somewhere, but related might be even better

raphaelm commented on tests/prefetch_related/tests.py:
```diff
@@ -256,6 +258,85 @@ def test_m2m_then_reverse_one_to_one_object_ids(self):
+            batch_size = connection.features.max_query_params - 25
+            batches = int(ceil(float(len(self.authors)) / batch_size))
+        else:
+            batches = 1
```
Good idea

raphaelm commented:
Hi @francoisfreitag, I changed my patch according to your comments. Happy for another review from you and/or @mjtamlyn (or anyone else)! :)

timgraham commented:
How about retrieving the connection like this `connection = connections[instances[0]._state.db]`. Isn't it  a safe assumption that all the instances are from the same db?

raphaelm commented:
> How about retrieving the connection like this connection = connections[instances[0]._state.db]. Isn't it a safe assumption that all the instances are from the same db?

Good idea. I changed that and rebased against current master.

timgraham commented:
Looks good. I added a commit with some edits if you could check it before I merge. I think this is more of a bug fix than a feature so I removed the release note.

raphaelm commented:
@timgraham Your commit looks good to me.

francoisfreitag commented on tests/prefetch_related/tests.py:
```diff
@@ -256,6 +259,81 @@ def test_m2m_then_reverse_one_to_one_object_ids(self):
+    SQLite. To improve test performance, a smaller value of max_query_params
+    is mocked.
+    """
+    test_size = 101
```
Maybe define a module constant, like `fake_max_query_param = 50` and use it in the `mock.patch` call and to define `TEST_SIZE=fake_max_query_param * 2 + 1`.

Also, I think 50 is too much, shouldn't 10 be enough?

timgraham replied:
I tried something similar but it turned out to be more annoying that I felt it was worth. I think the chances of someone changing one value and inadvertently not changing the other value is low.

As for what the test values should be... whatever works.

francoisfreitag commented on tests/prefetch_related/tests.py:
```diff
@@ -256,6 +259,81 @@ def test_m2m_then_reverse_one_to_one_object_ids(self):
+        if connection.features.max_query_params:
+            # 25 is the hardcoded safety margin to cope with additional
+            # variables in the query.
+            batch_size = connection.features.max_query_params - 25
```
I wonder if 25 should be defined in the [SQL constants module](https://github.com/django/django/blob/master/django/db/models/sql/constants.py)?

I am afraid changing 25 in the code might not be changed here, so the test would silently become obsolete.

timgraham commented on django/db/models/query.py:
```diff
@@ -1523,20 +1524,43 @@ def prefetch_one_level(instances, prefetcher, lookup, level):
+    else:
+        # Subtract a safety margin because the queryset might contain custom filtering
+        # expressions that also use variables.
+        batch_size = connection.features.max_query_params - 25
```
Would be fine I guess. I'm not too sure about the safety margin aspect in the first place, but it's SQLite so whatever works, I guess.

timgraham commented:
@francoisfreitag, feel free to send a PR to this branch with whatever edits you think will help. I'm in not rush to merge.

francoisfreitag commented:
It does not matter much, but fetching generic foreign key will be less efficient. The ORM tries to reduce the number of queries by [grouping similar content types together](https://github.com/django/django/blob/master/django/contrib/contenttypes/fields.py#L168-L209), however with batched queryset, that optimization cannot be as efficient (because only content of the same batch will be grouped together).

Splitting in batches can result in more query being triggered than expected. The instances for which objects will be prefetched [are grouped in a dict](https://github.com/django/django/blob/50ba8bc45ddb692edb50320d45aa7196faa42293/django/db/models/fields/related_descriptors.py#L561), a single query will be used to fetch the related objects, not matter how many times an instance in present in the original queryset. When fetching in batches, the same instance can trigger multiple prefetch request. An example may help clarify:
```python
# supposing max_query_params = 2 and no security padding.
amy = Author.objects.create(name='Amy')
belinda = Author.objects.create(name='Belinda')
book1 = Book.objects.create(author=amy)
book2 = Book.objects.create(author=amy)
book3 = Book.objects.create(author=amy)
book4 = Book.objects.create(author=belinda)

Book.objects.prefetch_related('author')
# Will select all (book.id, book.author_id):
# (1, 1)
# (2, 1)
# (3, 1)
# (4, 2)
```

LuRsT commented:
@francoisfreitag do you need any help with this PR?

francoisfreitag commented:
Hi @LuRsT,

I’m not actively working on this. To move forward, I think a test should be added to cover the issue described in https://github.com/django/django/pull/8309#issuecomment-309844445 and the code updated to pass the new tests.

This PR got off my radar, I’ll hopefully to give it a shot in the next weeks. Feel free to branch off that PR and suggest an improvement.

felixxm commented:
Closing due to inactivity.
